### PR TITLE
New private/public key pairs for broadcasting alert messages

### PR DESCRIPTION
**Pull request by bitcartel**
Implements #424 

Fixes and integrates method of sending alerts as described by upstream here:
- https://gist.github.com/laanwj/0e689cfa37b52bcbbb44

To send an alert:
- Copy private keys into alertkeys.h.
- Modify alert parameters and message found in sendalert.cpp
- Build and run to send the alert e.g. ./zcashd -printtoconsole -sendalert

Tested and verified with local nodes on alpha 6 testnet.

### CODE REVIEW

File: src/sendalert.cpp [hunk 1]
```diff
@@ -0,0 +1,201 @@
...
+    printf("vchSig=%s\n", HexStr(alert2.vchSig).c_str());
+
+    // Confirm
+    if (!mapArgs.count("-sendalert"))
+        return;
+    while (vNodes.size() < 1 && !ShutdownRequested())
+        MilliSleep(500);
+    if (ShutdownRequested())
+        return;
+#ifdef QT_GUI
```
Comment (ebfull) [inline src/sendalert.cpp]: Let's not have this, since we're not supporting the Qt client right now.

File: src/sendalert.cpp [hunk 2]
```diff
@@ -0,0 +1,201 @@
...
+
+    extern void ThreadSendAlert();
+
+  Add this call at the end of AppInit2:
+
+    threadGroup.create_thread(boost::bind(ThreadSendAlert));
+
+6. build bitcoind, then run it with -printalert or -sendalert
+  I usually run it like this:
+   ./bitcoind -printtoconsole -sendalert
```
Comment (daira) [inline src/sendalert.cpp]: ./zcashd

File: src/sendalert.cpp [hunk 3]
```diff
@@ -0,0 +1,201 @@
...
+5. Update init.cpp to launch the send alert thread. 
+  Define the thread function as external at the top of init.cpp:
+
+    extern void ThreadSendAlert();
+
+  Add this call at the end of AppInit2:
+
+    threadGroup.create_thread(boost::bind(ThreadSendAlert));
+
+6. build bitcoind, then run it with -printalert or -sendalert
```
Comment (daira) [inline src/sendalert.cpp]: Build zcashd

File: src/sendalert.cpp [hunk 4]
```diff
@@ -0,0 +1,201 @@
...
+    // filling to every node.
+    // After the relay time is past, new nodes are told about alerts
+    // when they connect to peers, until either nExpiration or
+    // the alert is cancelled by a newer alert.
+    // Nodes never save alerts to disk, they are in-memory-only.
+    //
+    CAlert alert;
+    alert.nRelayUntil   = GetTime() + 15 * 60;
+    alert.nExpiration   = GetTime() + 365 * 60 * 60;
+    alert.nID           = 1040;  // use https://en.bitcoin.it/wiki/Alerts to keep track of alert IDs
```
Comment (daira) [inline src/sendalert.cpp]: Update this number and URL.

File: src/sendalert.cpp [hunk 5]
```diff
@@ -0,0 +1,201 @@
...
+    alert.nCancel       = 0;   // cancels previous messages up to this ID number
+
+    // These versions are protocol versions
+    // 60002 : 0.7.*
+    // 70001 : 0.8.*
+    // 70002 : 0.9.*
+    alert.nMinVer       = 70002;
+    alert.nMaxVer       = 70002;
```
Comment (daira) [inline src/sendalert.cpp]: Do we need to be able to specify Zcash release numbers here?

**Comment by daira** [comment]
NACK, we need to consider how the versioning apples to Zcash.

File: src/sendalert.cpp (hunk 1 above)
Comment (bitcartel) [inline src/sendalert.cpp]: Ok

File: src/sendalert.cpp (hunk 3 above)
Comment (bitcartel) [inline src/sendalert.cpp]: Fixed comments.

File: src/sendalert.cpp (hunk 4 above)
Comment (bitcartel) [inline src/sendalert.cpp]: Ok, now set to 1000 and URL points to z.cash.  Can update with specific page when it exists.

**Comment by bitcartel** [comment]
If alerts are based on PROTOCOL_VERSION this would clearly imply that the alert system is to be used for consensus issues which alternative zcashd implementations should look out for.

If based on client release version, this would be seen as being application specific, so alternative implementations might want to have their own alert keys, which brings up questions such as (1) whether or not a node should relay alerts meant for other implementations (2) verifying those alerts requires some centralized management of the public keys for each implementation.

**Comment by daira** [comment]
Ah, I was confused by the protocol version seeming to be linked to a Bitcoin Core version, but it [hasn't been since Bitcoin 0.8.0](http://bitcoin.stackexchange.com/questions/13537/how-do-i-find-out-what-the-latest-protocol-version-is). So it's fine for us to increment the version independently for Zcash protocol changes starting from 70002.

File: src/sendalert.cpp [hunk 6]
```diff
@@ -103,13 +72,11 @@ void ThreadSendAlert()
     CAlert alert;
     alert.nRelayUntil   = GetTime() + 15 * 60;
     alert.nExpiration   = GetTime() + 365 * 60 * 60;
-    alert.nID           = 1040;  // use https://en.bitcoin.it/wiki/Alerts to keep track of alert IDs
+    alert.nID           = 1000;  // use https://z.cash to keep track of alert IDs
```
Comment (daira) [inline src/sendalert.cpp]: Make this https://github.com/zcash/zcash/wiki/specification#assigned-numbers

File: src/sendalert.cpp [hunk 7]
```diff
@@ -173,6 +140,7 @@ void ThreadSendAlert()
         MilliSleep(500);
     if (ShutdownRequested())
         return;
+#if 0
```
Comment (daira) [inline src/sendalert.cpp]: Let's just remove the code rather than ``#ifdef``ing it out.

**Comment by daira** [comment]
ACK modulo comments.

File: src/sendalert.cpp (hunk 6 above)
Comment (bitcartel) [inline src/sendalert.cpp]: Done

File: src/sendalert.cpp (hunk 7 above)
Comment (bitcartel) [inline src/sendalert.cpp]: Done

**Comment by daira** [comment]
ACK.

**Comment by ebfull** [comment]
I think that the `test/alert_tests.cpp` needs to be updated.

**Comment by bitcartel** [comment]
Ok, I'll take a look at the alert tests.

**Comment by bitcartel** [comment]
New test data generated.  Created a test fixture to make this easier in future.  Code and instructions on how to do this are in alert_tests.cpp.  Upstream has not released code for SignAndSave() so an alternate public implementation has been adapted for our purposes.

**Comment by ebfull** [comment]
ACK once @bitcartel sends us all the key!

**Comment by zookozcash** [comment]
Did I hear that upstream is deprecating the entire alerting mechanism? Would try potentially be interested in a contribution of this from us?

**Comment by ebfull** [comment]
Upstream removed the alert system from Bitcoin recently in its entirety. We still have the alert system (from our fork of 0.11.2) and this PR merely changes the key allowed to issue alerts.

**Comment by ebfull** [comment]
@zkbot r+

**Pull request merged**